### PR TITLE
[stdlib] - WIP replacing aspects of ForwardIndex with those from Collection

### PR DESCRIPTION
**Pull request by shawnce**
FYI @gribozavr @austinzheng 

Knocking out aspects of ForwardIndex updating affected code to use Collection equivalents.

Note in some places it isn’t yet possible or easy to fix the code so I dropped in fatalErrors("FIXME: swift-3-indexing-model") and/or “// FIXME: swift-3-indexing-model" comments.

Note many unit tests **fail** as a result of these changes because of the fatalErrors("FIXME: swift-3-indexing-model") added to unfinished parts of the code. This is WIP that I wanted to get some eyeballs on, up to you if you want to merge it into your branch or not.

### CODE REVIEW

**Comment by shawnce** [comment]
It should be noted that the code commented out in ForwardIndex is just temporary. It will be gone in a near future pass once things are fully off ForwardIndex. Then onto the other Index types.

File: stdlib/private/StdlibUnittest/StdlibUnittest.swift.gyb [hunk 1]
```diff
@@ -1835,12 +1835,14 @@ internal func _checkIncrementalAdvance<
   ${TRACE}
```
Comment (gribozavr) [inline stdlib/private/StdlibUnittest/StdlibUnittest.swift.gyb]: Seems like permissions were changed on this file.

Comment (shawnce) [inline stdlib/private/StdlibUnittest/StdlibUnittest.swift.gyb]: Good spot! Interesting. Xcode decide to change it on me? Will correct.

Comment (shawnce) [inline stdlib/private/StdlibUnittest/StdlibUnittest.swift.gyb]: fixed

**Comment by gribozavr** [comment]
I think you should be able to fix this FIXME in `Strideable` now:

```diff
public protocol Strideable : Comparable {
  // FIXME: We'd like to name this type "Distance" but for
  // <rdar://problem/17619038>
  /// A type that can represent the distance between two values of `Self`.
  associatedtype Stride : SignedNumber
```

**Comment by shawnce** [comment]
I will look at that as I move things from FowardIndex.Distance to Collection.IndexDistance (which is what I am currently working on).

File: stdlib/private/StdlibUnittest/StdlibUnittest.swift.gyb [hunk 2]
```diff
@@ -1835,12 +1835,14 @@ internal func _checkIncrementalAdvance<
...
   for i in instances {
-    let d = sign > 0 ? i.distance(to: limit) : -limit.distance(to: i)
+// TODO: swift-3-indexing-model - Review what we need to do here, unittest is focusing on wrong class now
+    let d: Instances.Iterator.Element.Distance = 0 //sign > 0 ? i.distance(to: limit) : -limit.distance(to: i)
 
     var offset: Instances.Iterator.Element.Distance = 0
     for _ in 0...(d * sign).toIntMax() {
-      let j = i.advanced(by: offset)
-      let k = i.advanced(by: offset + sign, limit: limit)
+// TODO: swift-3-indexing-model - Review what we need to do here, unittest is focusing on wrong class now
```
Comment (gribozavr) [inline stdlib/private/StdlibUnittest/StdlibUnittest.swift.gyb]: Right, this function needs to also receive a collection.

File: stdlib/public/core/Collection.swift [hunk 3]
```diff
@@ -242,9 +227,9 @@ public protocol Collection : Indexable, Sequence {
   ///   O(N) otherwise.
   var count: Index.Distance { get }
   
-  // The following requirement enables dispatching for indexOf when
-  // the element type is Equatable.
-  
+  /// The following requirement enables dispatching for indexOf when
+  /// the element type is Equatable.
+  ///
```
Comment (gribozavr) [inline stdlib/public/core/Collection.swift]: These three lines need to stay as a non-doc comment.

File: stdlib/public/core/ExistentialCollection.swift.gyb [hunk 4]
```diff
@@ -351,19 +351,28 @@ internal class _ForwardIndexBox<
   }
```
Comment (gribozavr) [inline stdlib/public/core/ExistentialCollection.swift.gyb]: Permissions were changed for this file, too.

File: stdlib/public/core/ExistentialCollection.swift.gyb [hunk 5]
```diff
@@ -351,19 +351,28 @@ internal class _ForwardIndexBox<
...
 
   func _distance(to other: _ForwardIndexBoxProtocol) -> AnyForwardIndex.Distance {
-    return numericCast(base.distance(to: unsafeUnbox(other)))
+// FIXME: swift-3-indexing-model - Review what we need to do here
```
Comment (gribozavr) [inline stdlib/public/core/ExistentialCollection.swift.gyb]: `Any*Collection` need to be almost completely rewritten now.

**Comment by gribozavr** [comment]
@shawnce Let me know if you need help with any FIXMEs.  I reviewed the patch and it looks reasonable.

File: stdlib/public/core/ExistentialCollection.swift.gyb (hunk 5 above)
Comment (shawnce) [inline stdlib/public/core/ExistentialCollection.swift.gyb]: Yeah I was starting to bang my head against how to do small changes to the Any*Collection code :) ... I will stub as possible to keep things happy until we can get around to working on it.

**Comment by shawnce** [comment]
@gribozavr For now failing unit tests should I mark them as expected to fail with a // "FIXME: swift-3-indexing-model" in them? If so what is the correct way to mark them as expecting to fail?

**Comment by gribozavr** [comment]
@shawnce You could mark them as `// XFAIL: *` and add a comment with the "swift-3-indexing-model" keyword to allow finding them easily.

File: stdlib/public/core/ExistentialCollection.swift.gyb (hunk 4 above)
Comment (shawnce) [inline stdlib/public/core/ExistentialCollection.swift.gyb]: fixed

**Comment by shawnce** [comment]
@gribozavr If you get the time can you merge the latest from your swift-3-api-guidelines into swift-3-indexing-model so I can pick up the more current code to work against. I could merge it into my fork directly but I think that makes things a little harder possibly when doing a PR back to you...?

File: stdlib/private/SwiftPrivate/SwiftPrivate.swift [hunk 6]
```diff
@@ -75,24 +75,27 @@ public func scatter<T>(a: [T], _ idx: [Int]) -> [T] {
...
+//    argsBuffer.append(0)
+//  }
+//
+//  return argsBuffer.withUnsafeBufferPointer {
+//    (argsBuffer) in
+//    let ptr = UnsafeMutablePointer<CChar>(argsBuffer.baseAddress)
+//    var cStrings = argsOffsets.map { ptr + $0 }
+//    cStrings[cStrings.count - 1] = nil
+//    return body(cStrings)
+//  }
```
Comment (dabrahams) [inline stdlib/private/SwiftPrivate/SwiftPrivate.swift]: To minimize churn, please comment code with `/*` … `*/` or `#if false`

Comment (shawnce) [inline stdlib/private/SwiftPrivate/SwiftPrivate.swift]: Noted!

**Comment by shawnce** [comment]
@gribozavr @dabrahams @moiseev 

Going to close out this WIP PR and likely squash the commits in my fork before opening a new PR at the next major stopping point in this work, so that it can be pulled into Apple's repo. ...unless I hear otherwise from folks.

**Comment by gribozavr** [comment]
SGTM.

**Pull request closed without merging**